### ISSUE TITLE
setExtraContext for individual errors

### ISSUE THREAD

**Issue by AdriVanHoudt** [opbeat/opbeat-angular#4]
Am I correct that `setExtraContext` only sets extra context for the next `captureException` or trace?
I want to log a bunch of extra info but adding to the error object does not seems to work...

atm doing 
```js
$opbeat.setExtraContext(rejection.config);
$opbeat.captureException(err);
```
but will this work and not bleed extra context on other `captureException`'s or traces?

**Comment by hmdhk** [comment]
Hi @AdriVanHoudt ,

Thanks for reaching out.
`setExtraContext` sets the context for all the errors happen after that.
Unfortunately at the moment there's no way to set extra context for individual errors.

I hope this answers your question.

**Comment by AdriVanHoudt** [comment]
hmm so I'd have to set the extra context, captureException and then set the extra context back to empty? This seems rather convoluted?

**Comment by hmdhk** [comment]
Unfortunately that wouldn't work, since the exception is sent asynchronously and just before sending we read the extra context. You need to wait for the promise that we return to resolve and then remove the context, at which point the context might not be the one that you original set for this error.

You're right, it is not meant to be used this way.
I'm going to change this to an improvement so we can plan for it.

Cheers,
Hamid

**Comment by AdriVanHoudt** [comment]
hmm yeah so that makes it kinda impossible :P

Just being able to set extra properties on the error object would solve my use case btw, thanks for taking this into the plan. Do you have an ETA on when you know more? (we have opbeat for our node servers already and would love to have the same power for logging on the client ^^)

**Comment by hmdhk** [comment]
Great to hear that you're using our node module :)
Our next planning session will be beginning of next week,
I will update this issue about the result.

**Comment by AdriVanHoudt** [comment]
@jahtalab do you have an update for me? I am really missing the ability to give more contextual data to the errors. I added the crossorigin attributes to hopefully get more info.

I'm getting 
```
SyntaxError: missing ) after argument list
(inline script)
```
With no stacktrace so no idea what is going on :/

**Comment by hmdhk** [comment]
@AdriVanHoudt , 

Sorry about the late response.
We have discussed it and this is something we want to provide but unfortunately it is still not on our planned timeline yet.

**Comment by AdriVanHoudt** [comment]
@jahtalab is it that hard to allow me to set custom properties on an error object?
I'm willing to make a PR for this, if your backend supports it obviously.

**Comment by hmdhk** [comment]
@AdriVanHoudt ,

It is not hard and our backend already supports it (it would be basically the same as setExtraContext that we already use).

The only thing is that we're refactoring this part and there might be some changes (This is one of the reasons we couldn't plan this feature).
The refactoring should be (partly) done this week though.

**Comment by AdriVanHoudt** [comment]
@jahtalab ok can you let me know when you are ready with the refactoring then? We can either look together at a solution then and/or I can do a PR for it

**Comment by hmdhk** [comment]
@AdriVanHoudt , 
Sure thing 👍

**Comment by AdriVanHoudt** [comment]
@jahtalab I will be getting back to you next week then to check in ;)

**Comment by AdriVanHoudt** [comment]
I realise that probably nothing is going to happen in the last week of the year ^^ so let me check in next week :D

Happy holidays 🎆

**Comment by hmdhk** [comment]
@AdriVanHoudt , The feature is coming up. 
Once it is published you can use it like this:

```javascript
     try {
       throw new Error('error')
     } catch (error) {
      error['_opbeat_extra_context'] = {context: 'test'}
      _opbeat('captureException', error)
     }
```

Let me know if this is suitable in your case.

Happy holidays

**Comment by AdriVanHoudt** [comment]
`error['_opbeat_extra_context']` can be an object containing anything right?

**Comment by hmdhk** [comment]
@AdriVanHoudt , That's right.

**Comment by AdriVanHoudt** [comment]
@jahtalab sounds good to me!

**Comment by hmdhk** [comment]
@AdriVanHoudt , 

We have decided to make the interface the same as our NodeJS integration, so instead you can set any property on the error and it will be sent up as extra context. The only limitation is that you can not set objects, only basic types (string, number and dates).
Does this work for you?

**Comment by AdriVanHoudt** [comment]
@jahtalab even better to make them the same ^^, I think the node module allows objects but stringifies them before sending

**Comment by hmdhk** [comment]
@AdriVanHoudt ,
I have release opbeat-angular@3.9.1 which has this feature.
you can add contextual data to errors like this:


```javascript
     var error = new Error('error')
     error.testContext = 'test'

     _opbeat('captureException', error) 
     // or just throw error;
```

Properties containing Objects/functions are ignored (to avoid recursive or big objects). But you can assign `Date` object which will be converted to ISO string.

Note: The contextual data on the error will be merged with the contextual data you set using `$opbeat.setExtraContext`.

**Comment by AdriVanHoudt** [comment]
@jahtalab nice! 🎉 Why not stringify objects?

**Comment by hmdhk** [comment]
@AdriVanHoudt , it is to avoid unexpectedly sending big objects or to stringify recursive objects. For example if other libraries set objects on the error object. However We might change this in the future to include objects as well.

**Comment by AdriVanHoudt** [comment]
Does the node lib not do this? cc @watson

**Comment by hmdhk** [comment]
@AdriVanHoudt , This is also the same in our Node module.

**Comment by AdriVanHoudt** [comment]
hmm I definitely have things in the extra tab that are stringified objects ![](https://puu.sh/togA1/f01b7c8405.png)

**Comment by watson** [comment]
**TL;DR:**

The Node agent doesn't stringify the `extra` data - it's just how it's rendered on opbeat.com.

**Long answer:**

The Node agent will take anything you've added as an `extra` property and send it almost as-is. The only caveat are things that can't be converted to JSON - like circular references, functions etc. But those should not be added as `extra` properties anyway 😉

The Node agent will also look for properties on the `Error` object it self. Those will be added automatically as `extra` properties. This is useful when you for instance are logging an error that comes from node core as they in many cases contain error codes and other useful information as properties.

**Gotcha regarding properties on `Error` objects:**

The Node agent will ignore properties on the `Error` object of type `object` (except for Dates which is converted to a string). This is because some modules/frameworks will add non-data objects to the error which it doesn't make sense to log (like a reference to the HTTP request object or something). So if you want to log an actual object (like the one from the screenshot above), it should be added as a regular `extra` property as I suspect you also do 😃

**Comment by hmdhk** [comment]
We can provide other ways to capture objects as contextual data,
There are a couple of options:

1. You would call `_opbeat.captureException(error, {extra:'test'})` after catching the error (you can't rethrow the error)
2. You would set a predefined property on the error after catching it and either use `_opbeat.captureException` or just rethrow it

Of course feel free to suggest if none of the above works for you.

**Comment by AdriVanHoudt** [comment]
@watson thanks for explaining, makes sense now
@jahtalab the closer to node the better, I see that in my code I'm doing `opbeat.captureError(error,{ extra: extraStuff, logger: loggerName })` so option 1 looks good!

**Comment by hmdhk** [comment]
@AdriVanHoudt ,

You can now send extra data for errors (version 3.10.0).

Example: 

```
    $opbeat.captureException(error, {
      extra: {
        foo: 'bar'
      }
    })

```

Let me know how it works out for you.

**Comment by AdriVanHoudt** [comment]
I will come back to you when I've implemented this!

**Comment by AdriVanHoudt** [comment]
seems to work fine! 🎉 
@jahtalab thanks for the effort!
